### ISSUE TITLE
bug: relative navigation doesn't work inside route guards

### ISSUE THREAD

**Issue by jorroll** [angular/angular#22763]
## I'm submitting a...

<pre><code>
[x] Bug report  
</code></pre>

## Current behavior

Inside a route guard, injecting `ActivatedRoute` and attempting relative navigation via `this.router.navigate(['../sibling'], {relativeTo: this.route})` results in a `Error: Cannot match any routes. URL Segment: 'sibling'`

## Expected behavior

Relative navigation should work normally

## Minimal reproduction of the problem with instructions

A stackblitz example of the issue can be seen here: https://stackblitz.com/edit/angular-gitter-k7fjds

## What is the motivation / use case for changing the behavior?

1. This appears to be a bug
2. If a user navigates to a child route and includes a specific queryParam, I'd like to redirect the user to a sibling route. Currently, this does not appear to be possible inside a route guard with relative navigation. 

## Additional Information

It appears that the `ActivatedRoute` injected into a route guard is always an instance of the root route. Inside a route guard, `this.router.navigate(['.'], {relativeTo: this.route})` always takes the user back to the root route, rather than navigating them to the same route they are currently on.

## Environment

<pre><code>
Angular version: 5.0.0

Browser:
- [x] Chrome (desktop) version 64

**Comment by jasonaden** [comment]
Looks related to #17023

**Comment by thefliik** [comment]
This issue appears to be more nuanced then I originally thought. Before posting, I had noticed that the injected `ActivatedRoute` appeared to be the root route, so I tried looping through the `ActivatedRoute` and then using the result for relative navigation, like so
```
constructor(
  private router: Router,
  private route: ActivatedRoute,
) {}

canActivate(next: ActivatedRouteSnapshot, state: RouterStateSnapshot): boolean {
  const emailAddress = next.queryParams['emailAddress'] as string | undefined;

  let route = this.route;

  while (route.firstChild) {route = route.firstChild}

  if (emailAddress) {
    this.router.navigate(['../sibling'], {relativeTo: route});
  }

  return true;
}
```
This didn't work however, so I created this issue.

Just fooling around now though, I see this DOES work when navigating from the `/parent/sibling` route to the `/parent/child` route. It DOES NOT work when navigating from `/parent` to `/parent/child`.

**Comment by thefliik** [comment]
The issue seems to be that the _current_ `ActivatedRoute` is being injected and used (which makes sense, seeing as it _is_ the currently activated route), and there isn't a way to get what will _become_ the active route in an `ActivatedRoute` format. (the next route is only provided as an `ActivatedRouteSnapshot`, which can't be used for relative navigation).

**Comment by thefliik** [comment]
Ya, my current workaround is to fiddle with the raw url.

**Comment by SamVerschueren** [comment]
I wrote this small utility function that I use in my guard

```ts
private parseUrl(url: string, redirectTo: string) {
	const urlTokens = url.split('/');
	const redirectToTokens = redirectTo.split('/');

	let token = redirectToTokens.shift();

	while (token) {
		if (token !== '.' && token !== '..') {
			redirectToTokens.unshift(token);
			break;
		}

		if (token === '..') {
			urlTokens.pop();
		}

		token = redirectToTokens.shift();
	}

	urlTokens.push(...redirectToTokens);

	return urlTokens.join('/');
}
```

You can call it the following ways

```ts
parseUrl('/foo/bar/baz', './qux');
//=> /foo/bar/baz/qux

parseUrl('/foo/bar/baz', '../qux');
//=> /foo/bar/qux

parseUrl('/foo/bar/baz', '../../qux');
//=> /foo/qux
```

The first argument that I pass in is `state.url` because that's the route I want to go to.

I might be missing some edge cases but it works for me, so thought I could share it.

**Comment by mantuonweb** [comment]
I am facing the same issue relative routing does not working in guard services?
Guys did you get any fix

**Comment by thefliik** [comment]
@mantuonweb you didn't actually read this thread did you?

**Comment by mantuonweb** [comment]
Yes , I did but may be this fixed in updated one.
Just wants to know that

**Comment by MarcusRiemer** [comment]
It really is a pity that this is so hard. Another workaround was mentioned in #28682, but this also doesn't seem to work anymore with Angular 9: 

```
return (this.router.parseUrl(`${state.url}/../unknown`));
```

For me the "expected" solution would be to accept a `RouterStateSnapshot` in `Router.createUrlTree`. From what I can tell `snapshot` property is the only thing from the route that is used anyway: https://github.com/angular/angular/blob/5edeee69dd0a7ffff0790d6f817c95f58914996d/packages/router/src/create_url_tree.ts#L145-L158

Called from https://github.com/angular/angular/blob/5edeee69dd0a7ffff0790d6f817c95f58914996d/packages/router/src/create_url_tree.ts#L27

Which in turn is called from https://github.com/angular/angular/blob/9.0.2/packages/router/src/router.ts#L951

@jasonaden (or anyone from the Angular team): This seems like a trivial fix, so I guess im am overlooking something?

**Comment by NLueg** [comment]
Still having this Issue in Angular 9 ... It's a pity, that there is no priority in fixing this.

**Comment by ddralves** [comment]
Any update on this issue?

**Comment by sergeypayu** [comment]
Here is my workaround:

```
    canActivate(route: ActivatedRouteSnapshot): UrlTree | boolean {
            // construct full url from urlSegments descending from root
            const urlSegments = route.pathFromRoot
                .map((snapshot) => snapshot.url)
                .reduce((acc, urlSegments) => acc.concat(urlSegments), [])
                .filter((urlSegment) => !!urlSegment);
            const currentUrl = urlSegments.map((urlSegment) => urlSegment.path);
            const relativeUrl = currentUrl.concat(['page']);
            return this.router.createUrlTree(relativeUrl);
    }
```

**Comment by atscott** [comment]
The TL;DR for this one is that this is something I would absolutely love to do but will be a breaking change to update `router.createUrlTree` to do it. 

The implementation for creating a URL tree is _really_ confusing and actually requires that the `ActivatedRoute/ActivatedRouteSnapshot` be part of the current router url. In reality, we should really be able to create a url from any `ActivatedRouteSnapshot`, but that requires the shape of the `ActivatedRoute`/`ActivatedRouteSnapshot` to be valid (which is commonly not the case in tests due to stubbing this out).

There is maybe an opportunity to introduce this new way of creating the URL as a standalone function exported as public API in the `router` and then later making the breaking change to include it in `router.createUrlTree`. This would be problematic for discovery and documentation but would at least provide an early opt-in for anyone trying to do this.
